tikv-client/src/main/java/com/pingcap/tikv/codec/RowDecoderV2.java

```diff
@@ -0,0 +1,180 @@
+        return decodeTimestamp(colData, DateTimeZone.UTC);
+      case TypeDuration:
+      case TypeYear:
+        return decodeInt(colData);
+      case TypeEnum:
+        return decodeEnum(colData, tp.getElems());
+      case TypeSet:
+        return decodeSet(colData, tp.getElems());
+      case TypeJSON:
+        return decodeJson(colData);
+      case TypeNull:
+        return null;
+      case TypeDecimal:
+      case TypeGeometry:
+      case TypeNewDate:
+        throw new CodecException("type should not appear in colData");
+      default:
+        throw new CodecException("invalid data type " + tp.getType().name());
+    }
+  }
+
+  private static long decodeInt(byte[] val) {
+    switch (val.length) {
+      case 1:
+        return val[0];
+      case 2:
+        return new CodecDataInputLittleEndian(val).readShort();
+      case 4:
+        return new CodecDataInputLittleEndian(val).readInt();
+      default:
+        return new CodecDataInputLittleEndian(val).readLong();
+    }
+  }
+
+  private static float decodeFloat(byte[] val) {
+    return (float) decodeDouble(val);
+  }
+
+  private static double decodeDouble(byte[] val) {
+    CodecDataInput cdi = new CodecDataInput(val);
+    if (val.length < 8) {
+      throw new CodecException("insufficient bytes to decode value");
+    }
+    long u = cdi.readLong();
+    // signMask is less than zero in int64.
+    if ((u & SIGN_MASK) < 0) {
+      u &= ~SIGN_MASK;
+    } else {
+      u = ~u;
+    }
+    return Double.longBitsToDouble(u);
+  }
+
+  private static BigDecimal decodeDecimal(byte[] val) {
+    return DecimalCodec.readDecimal(new CodecDataInputLittleEndian(val));
+  }
+
+  private static byte[] trimLeadingZeroBytes(byte[] bytes) {
+    if (bytes.length == 0) {
+      return bytes;
+    }
+    int pos = 0, posMax = bytes.length - 1;
+    for (; pos < posMax; pos++) {
+      if (bytes[pos] != 0) {
+        break;
+      }
+    }
+    return Arrays.copyOfRange(bytes, pos, bytes.length);
+  }
+
+  private static byte[] decodeBit(long val, int byteSize) {
+    if (byteSize != -1 && (byteSize < 1 || byteSize > 8)) {
+      throw new CodecException("Invalid byteSize " + byteSize);
+    }
+    CodecDataOutput cdo = new CodecDataOutput();
+    cdo.writeLong(val);
+    if (byteSize != -1) {
+      return trimLeadingZeroBytes(cdo.toBytes());
+    } else {
+      return Arrays.copyOfRange(cdo.toBytes(), 8 - byteSize, 8);
+    }
+  }
+
+  private static Timestamp decodeTimestamp(byte[] val, DateTimeZone tz) {
+    ExtendedDateTime extendedDateTime =
+        DateTimeCodec.fromPackedLong(new CodecDataInputLittleEndian(val).readLong(), tz);
+    // Even though null is filtered out but data like 0000-00-00 exists
+    // according to MySQL JDBC behavior, it can chose the **ROUND** behavior converted to the
+    // nearest
+    // value which is 0001-01-01.
+    if (extendedDateTime == null) {
+      return DateTimeCodec.createExtendedDateTime(tz, 1, 1, 1, 0, 0, 0, 0).toTimeStamp();
```

Comment:
Do we have similar round behavior in default row decoder?

Reply:
See https://github.com/pingcap/tispark/blob/316d552a0a742b3da93e69f41e115b50908dfbfb/tikv-client/src/main/java/com/pingcap/tikv/types/AbstractDateTimeType.java#L59-L70